File: solutions/free-pascal/binary-search/1/BinarySearch.pas
unit BinarySearch;

{$mode ObjFPC}{$H+}

interface

type
  TExtendedArray = Array Of Extended;

function find(const AArray: TExtendedArray; const AValue : Extended) : Integer;

implementation

uses SysUtils;

function find(const AArray: TExtendedArray; const AValue : Extended) : Integer;
var 
  left, right, mid : integer;
  val : Extended;
begin
  left := 0;
  right := length(AArray) - 1;
 
  while left<=right do
  begin
    mid := (right+left) div 2;
    val := AArray[mid];
    if (val=AValue) then
      exit(mid)
    else if (val>AValue) then
      right := mid - 1
    else
      left := mid + 1;
  end;

  raise Exception.Create('value not in array');

end;

end.

File: solutions/dart/binary-search/1/lib/binary_search.dart
import 'package:binary_search/value_not_found_exception.dart';

class BinarySearch {
  late List<int> _data;

  BinarySearch(List<int> data){
    _data = List.from(data);
  }

  int find(int value){
    var left=0;
    var right=_data.length-1;
    while(left<=right){
      var mid = (left + right)~/2;
      if(_data[mid] == value)
        return mid;
      else if(_data[mid]>value)
        right = mid-1;
      else
        left = mid+1;
    }
    throw ValueNotFoundException("not found");
  }
}
